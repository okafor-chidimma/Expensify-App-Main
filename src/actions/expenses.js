import database from '../firebase/firebase';
// Add expense
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

// when the function is called, it returns a function which only works with the help of thunk middleware.
// Then the component dispatches this function to the redux store
// the redux store executes this function internally and every other thing inside the function
export const startAddExpense = (expense = {}) => {
  // this return function is internally ran by redux and it gets called with dispatch as an arg
  return dispatch => {
    const { amount = 0, createdAt = 0, note = '', description = '' } = expense;
    const expenseObj = { amount, description, createdAt, note };

    // to return a promise so we can use jest to test to make sure it does what it is meant to do
    return database
      .ref('expenses')
      .push(expenseObj)
      .then(ref => {
        // the first dispatch from the component just makes this function available to redux store
        // it is this dispatch that changes the redux store
        dispatch(
          addExpense({
            id: ref.key,
            ...expenseObj
          })
        );
      })
      .catch(error => {
        console.log('could not insert into db', error);
      });
  };
};

// remove expense

export const removeExpense = ({ id }) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  };
};

// Edit Expense

export const editExpense = ({ id }, updatedExpense) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updatedExpense
  };
};

// To Set Expense That is already in the db

// expenses here is an array
export const setExpense = expenses => ({
  type: 'SET_EXPENSE',
  expenses
});

export const startSetExpense = () => {
  // this return function is internally ran by redux and it gets called with dispatch as an arg
  return dispatch => {
    // to return a promise so we can use jest to test to make sure it does what it is meant to do
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpense(expenses));
      })
      .catch(error => {
        console.log('could not read from db', error);
      });
  };
};
