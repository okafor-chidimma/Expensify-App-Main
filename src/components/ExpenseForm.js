import React, { Component } from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense
      ? (this.props.expense.amount / 100).toString()
      : '',
    createdAt: this.props.expense
      ? moment(this.props.expense.createdAt)
      : moment(),
    calendarFocused: false,
    error: ''
  };
  handleAddDescription = event => {
    const description = event.target.value;
    this.setState(() => ({
      description
    }));
  };
  handleAddNote = event => {
    const note = event.target.value;
    this.setState(() => ({
      note
    }));
  };
  handleAddAmount = event => {
    const amount = event.target.value;
    // this condition is for users to be able to clear their input
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    } else {
      console.log('invalid amount');
      this.setState(() => ({ error: 'Invalid Input For Amount' }));
    }
  };
  onDateChange = date => {
    if (date) {
      this.setState({
        createdAt: date
      });
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState({
      calendarFocused: focused
    });
  };
  onFormSubmit = event => {
    event.preventDefault();
    if (!this.state.description || !this.state.amount) {
      console.log('i should have error');
      this.setState(() => ({
        error: 'Please neither amount nor description can be empty'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      console.log('submitted');
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  render() {
    return (
      <form onSubmit={this.onFormSubmit} className='form'>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          type='text'
          name='description'
          className='text-input'
          placeholder='Description'
          autoFocus
          value={this.state.description}
          onChange={this.handleAddDescription}
        />
        <input
          type='text'
          className='text-input'
          placeholder='amount'
          value={this.state.amount}
          onChange={this.handleAddAmount}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className='textarea'
          placeholder='Enter a note for the expense'
          value={this.state.note}
          onChange={this.handleAddNote}
        ></textarea>
        <div>
          <button className='button'>
            {this.props.buttonState === 'edit' ? 'Edit Expense' : 'Add Expense'}
          </button>
        </div>
      </form>
    );
  }
}

export default ExpenseForm;
