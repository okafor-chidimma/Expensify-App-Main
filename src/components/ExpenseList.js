import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpense from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
  <div className='content-container'>
    <div className='list-header'>
      <div className='show-for-mobile'>Expenses</div>
      <div className='show-for-desktop'>Expenses</div>
      <div className='show-for-desktop'>Amount</div>
    </div>
    <div className='list-body'>
      {expenses.length === 0 ? (
        <div className='list-item list-item--message'>
          <span>No expenses</span>
        </div>
      ) : (
        expenses.map(expense => {
          return <ExpenseListItem {...expense} key={expense.id} />;
        })
      )}
    </div>
    {
      // props.expenses.map(({ description, amount, createdAt, id }) => {
      //   return (
      //     <ExpenseListItem
      //       description={description}
      //       amount={amount}
      //       createdAt={createdAt}
      //       key={id}
      //     />
      //   )
      // })
    }
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpense(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);

// it can re-written as

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses,
//     filters: state.filters

//   }
// })(ExpenseList);
// export default ConnectedExpenseList;
