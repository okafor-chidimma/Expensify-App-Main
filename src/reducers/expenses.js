// Expenses Reducer
const expenseStateDefault = [];
const expenseReducer = (state = expenseStateDefault, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updatedExpense
          }
        } else {
          return expense
        }
      });
    default:
      return state;
  }
};

export default expenseReducer;