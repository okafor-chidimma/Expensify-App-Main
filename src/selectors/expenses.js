import moment from 'moment';

// to sort the expense and return data that matches
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
    // this includes will always be true as long as the text is an empty string or a string contained in description
    // i.e textMatch will only be false is the text is not empty and cannot be found in description
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // console.log(textMatch, 'text match');
    // console.log(endDate, 'end date');
    // console.log(createdAtMoment, 'created at moment');
    // console.log(expense.createdAt, 'value of');
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  });

}

export default getVisibleExpenses;