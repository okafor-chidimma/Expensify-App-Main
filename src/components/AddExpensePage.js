import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
export class AddExpensePage extends React.Component {
  onSubmit = expenseDetails => {
    console.log(expenseDetails, 'i am expense details');
    this.props.startAddExpense(expenseDetails);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense Page</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm buttonState='create' onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}
const mapDispatchToProps = dispatch => ({
  // this dispatch sends the function to the store
  startAddExpense: expense => dispatch(startAddExpense(expense))
});
//             connect(mapStateToProps,mapDispatchToProps)
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
