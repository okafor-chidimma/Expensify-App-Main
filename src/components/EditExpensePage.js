import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = expenseDetails => {
    console.log(expenseDetails, 'i am expense details');
    // dispatch(editExpense({ id: props.match.params.id }, expense))
    this.props.startEditExpense(this.props.match.params.id, expenseDetails);
    this.props.history.push('/');
  };
  onButtonClick = () => {
    this.props.startRemoveExpense(this.props.match.params.id);
    // dispatch(removeExpense({ id: props.match.params.id }))
    this.props.history.push('/');
  };
  render() {
    // console.log(this.props);

    return (
      <div>
        Editing the expense of {this.props.match.params.id}
        <ExpenseForm
          expense={this.props.expense}
          buttonState='edit'
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onButtonClick}>Remove Item</button>
      </div>
    );
  }
}

// connect method has access to the connected component props by accepting a second parameter which is the props

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};
const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
