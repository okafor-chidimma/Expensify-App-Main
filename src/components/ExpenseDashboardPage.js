import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseSummary from "./ExpenseSummary";
const ExpenseDashBoardPage = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashBoardPage;
