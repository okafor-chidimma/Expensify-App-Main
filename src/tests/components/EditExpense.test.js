import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper, match;
describe('should render the <EditExpensePage/>', () => {
  // declare the functions the component is expecting as props
  beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    match = { params: { id: expenses[0].id } }
    wrapper = shallow(<EditExpensePage history={history} editExpense={editExpense} removeExpense={removeExpense} expense={expenses[0]} match={match} />);

  });
  it('should render <EditExpensePage/> correctly with right props', () => {
    /*
      when you simulate an event, it is the same as the event occurring so that means that the call back function executes but when you just have a function that is not being triggered by an event, you create a mock of the function using jest and assert what you want to assert
    */
    expect(wrapper).toMatchSnapshot();

  });
  it('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0])
  });
  it('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
  })
})