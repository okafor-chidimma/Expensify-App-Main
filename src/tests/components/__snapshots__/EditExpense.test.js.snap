// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the <EditExpensePage/> should render <EditExpensePage/> correctly with right props 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense Page
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      buttonState="edit"
      expense={
        Object {
          "amount": 54500,
          "createdAt": 0,
          "description": "January rent",
          "id": "1",
          "note": "Last rent before moving",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
